Added module Test.Shelley.Spec.Ledger.Generator.Scripts

TimSheard opened the pull request:
This module collects together all the operations that operate on the type family (Core.Script era) to write property tests.
The ScriptClass is made a superclass of GenEra
  class ScriptClass era => EraGen era where ...
Small changes are made to 7 files, to call these generic Script operations.

TimSheard committed:
Added module Test.Shelley.Spec.Ledger.Generator.Scripts

This module collects together all the operations that operate on the type family (Core.Script era).
The ScriptClass is made a superclass of GenEra
  class ScriptClass era => EraGen era where ...
Small changes are made to 7 files, to call these generic Script operations.

TimSheard committed:
Added module Test.Shelley.Spec.Ledger.Generator.Scripts

This module collects together all the operations that operate on the type family (Core.Script era).
The ScriptClass is made a superclass of GenEra
  class ScriptClass era => EraGen era where ...
Small changes are made to 7 files, to call these generic Script operations.

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Scripts.hs:
```diff
@@ -22,8 +22,8 @@ module Shelley.Spec.Ledger.Scripts
     ScriptHash (..),
-    getKeyCombination,
-    getKeyCombinations,
+    -- getKeyCombination,  GONE
```
Just delete in such a case!

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Scripts.hs:
```diff
@@ -186,6 +187,7 @@ getKeyCombinations (RequireAnyOf msigs) = List.concatMap getKeyCombinations msig
 getKeyCombinations (RequireMOf m msigs) =
   let perms = map (take m) $ List.permutations msigs
    in map (concat . List.concatMap getKeyCombinations) perms
+-}
```
:+1: Good to get rid of this!

nc6 commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Scripts.hs:
```diff
@@ -0,0 +1,266 @@
+--   used in the 'quantify' and 'unQuantify' methods of ScriptClass.
+data Quantifier t = AllOf [t] | AnyOf [t] | MOf Int [t] | Leaf t
+
+anyOf :: forall era. ScriptClass era => Proxy era -> [Script era] -> Script era
```
Given that we already have `AllowAmbiguousTypes` enabled, is there much point in having the proxy here? We can just resolve this with type annotations, without the extra argument.

nc6 commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Scripts.hs:
```diff
@@ -0,0 +1,266 @@
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE PatternGuards #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
```
Please run the formatter, just to make later diffs smaller.

At some point we should run it over all of the tests (I guess I missed enabling it there during some move or another), but in the meantime it will help to avoid changes later, especially for new files.

nc6 reviewed (changes requested):
Looks great, only please remove the code rather than commenting it out!

Let's get #1997 merged first, since this sits on top if it

uroboros commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Core.hs:
```diff
@@ -220,9 +224,9 @@ class
     Gen (Core.TxBody era)
 
-  eraScriptWitnesses :: Core.Script era -> [[KeyHash 'Witness (Crypto era)]]
+  -- eraScriptWitnesses :: Core.Script era -> [[KeyHash 'Witness (Crypto era)]] GONE
```
it's nice to see these go!

uroboros commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Core.hs:
```diff
@@ -272,7 +276,7 @@ data KeySpace era = KeySpace_
 
 pattern KeySpace ::
-  EraGen era =>
+  ScriptClass era =>
```
this is a much more appropriate constraint than what I had with EraGen

uroboros commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Scripts.hs:
```diff
@@ -0,0 +1,266 @@
+  quantify :: Proxy era -> Script era -> Quantifier (Script era)
+  unQuantify:: Proxy era -> Quantifier (Script era) -> Script era
+
+type ScriptPairs era = [(Core.Script era, Core.Script era)]
```
I've just recently gotten rid of `MultiSigPairs` - I vote for skipping the type synonym here

uroboros commented on shelley-ma/shelley-ma-test/src/Test/Cardano/Ledger/EraBuffet.hs:
```diff
@@ -163,11 +168,61 @@ instance FromCBOR (Annotator TestTxBody) where fromCBOR = pure <$> fromCBOR
+-- EraGen instaces for the MaryEra
+-- =====================================
+
+instance (CryptoClass.Crypto c) => ScriptClass (MaryEra c) where
```
These Mary and Allegra instances can move to the `shelley-ma-test package` (which references this one and can share what it needs to).

uroboros committed:
Remove value generator arg in benchmarking

TimSheard committed:
Adding the classes ValueClass and TxBodyClass

Generalizing genesisId and genUtxo0
Added ValueClass instances for Mary and Allegra Eras.

TimSheard committed:
Completed EraGen instances for MaryEra and AllegraEra.

These appear in the Modules  Test.Cardano.Ledger.Mary and  Test.Cardano.Ledger.Allegra
which are in the same directory as the EraBuffet:  Test.Cardano.Ledger.EraBuffet
Had to generalise:  genValue:: [ScriptHash era] -> Integer -> Integer -> Gen(Core.Value era)

TimSheard committed:
renamed Test.Shelley.Spec.Ledger.Generator.Scripts.

In order to better point out that this module has classes for all the type families
we rename it to Test.Shelley.Spec.Ledger.Generator.TypeFamilyClasses. We also generalized
genEraTxBody to take a list of script Hashes in addition to its other arguments.
Orlolised.

uroboros commented:
This work has been moved to this PR -> https://github.com/input-output-hk/cardano-ledger-specs/pull/2012
